### PR TITLE
Identify gradle project directories

### PR DESCRIPTION
**Pull request by kotarak**
As needed for [gradle support in `cider-jack-in`](https://github.com/clojure-emacs/cider/pull/1380).

### CODE REVIEW

File: clojure-mode.el [hunk 1]
```diff
@@ -1041,16 +1041,13 @@ nil."
...
-               (t (if (file-in-directory-p lein-project-dir boot-project-dir)
-                      lein-project-dir
-                    boot-project-dir))))))))
+  (let* ((dir-name (or dir-name default-directory))
+         (choices (delq nil
+                        (mapcar (lambda (fname)
+                                  (locate-dominating-file dir-name fname))
+                                '("project.clj" "build.boot" "build.gradle")))))
+    (when (< 0 (length choices))
+      (car (sort choices (lambda (a b) (file-in-directory-p a b)))))))
```
Comment (Malabarba) [inline clojure-mode.el]: aka `#'file-in-directory-p` :-)

Comment (kotarak) [inline clojure-mode.el]: *clonk* *headhitstable*

Indeed.

File: clojure-mode.el [hunk 2]
```diff
@@ -1041,16 +1041,13 @@ nil."
...
-               ((not boot-project-dir) lein-project-dir)
-               (t (if (file-in-directory-p lein-project-dir boot-project-dir)
-                      lein-project-dir
-                    boot-project-dir))))))))
+  (let* ((dir-name (or dir-name default-directory))
+         (choices (delq nil
+                        (mapcar (lambda (fname)
+                                  (locate-dominating-file dir-name fname))
+                                '("project.clj" "build.boot" "build.gradle")))))
+    (when (< 0 (length choices))
```
Comment (bbatsov) [inline clojure-mode.el]: I'd normally write this as `(> (length choices) 0)` as it reads more naturally to me.

File: clojure-mode.el [hunk 3]
```diff
@@ -1041,16 +1041,13 @@ nil."
...
+      (car (sort choices 'file-in-directory-p)))))
```
Comment (bbatsov) [inline clojure-mode.el]: `#'file-in-directory-p` -> we'll get more assistance from the compiler this way

File: clojure-mode.el [hunk 4]
```diff
@@ -1041,16 +1041,13 @@ nil."
...
-         (cond ((not lein-project-dir) boot-project-dir)
-               ((not boot-project-dir) lein-project-dir)
-               (t (if (file-in-directory-p lein-project-dir boot-project-dir)
-                      lein-project-dir
-                    boot-project-dir))))))))
+  (let* ((dir-name (or dir-name default-directory))
+         (choices (delq nil
+                        (mapcar (lambda (fname)
+                                  (locate-dominating-file dir-name fname))
+                                '("project.clj" "build.boot" "build.gradle")))))
```
Comment (bbatsov) [inline clojure-mode.el]: I'd extract those to a defcustom.

**Comment by bbatsov** [comment]
Apart from my small remarks - squash the commits and update the changelog.

File: clojure-mode.el (hunk 3 above)
Comment (kotarak) [inline clojure-mode.el]: Ah. Sorry. Didn't know that #' was a thing in emacs. Thinking too much clojure.

**Comment by kotarak** [comment]
Messed up squashing.

**Pull request closed without merging**